File: 2024/20/main.py
#!/usr/bin/env python3

import sys
import heapq
import tqdm
from collections import deque

DIRS = [(-1,0),(0,1),(1,0),(0,-1)]

def doPart1(): # pypy3 -> 1:11
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    GRID = open(filename).read().strip().split("\n")

    R = len(GRID)
    C = len(GRID[0])

    for r in range(R):
        for c in range(C):
            if GRID[r][c] == "E":
                er,ec = r,c
            if GRID[r][c] == "S":
                sr,sc = r,c

    def runOne(cheatR=-1, cheatC=-1):
        Q = []
        SEEN = set()
        heapq.heappush(Q, (0,sr,sc,1))
        DIST = {}
        best = None
        while Q:
            d,r,c,dir = heapq.heappop(Q)
            if (r,c,dir) not in DIST:
                DIST[(r,c,dir)] = d
            if r==er and c==ec and best is None:
                best = d
            if (r,c,dir) in SEEN:
                continue
            SEEN.add((r,c,dir))
            for k in [0,1,3]:
                dr,dc = DIRS[(dir+k)%4]
                rr,cc = r+dr,c+dc
                if GRID[rr][cc] != '#' or (rr==cheatR and cc==cheatC):
                    heapq.heappush(Q, (d+1, rr,cc,(dir+k)%4))
        return best

    noCheat = runOne()
    ans = 0
    for cr in tqdm.tqdm(range(1,(R-1))):
        for cc in range(1,(C-1)):
            newTime = runOne(cr,cc)
            if noCheat - newTime >= 100:
                ans+=1
    print(ans)

def doPart1Better():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    GRID = open(filename).read().strip().split("\n")

    R = len(GRID)
    C = len(GRID[0])

    for r in range(R):
        for c in range(C):
            if GRID[r][c] == "S":
                sr,sc = r,c

    def runOne():
        DIST = {}
        q=deque()
        q.append((sr,sc))
        DIST[(sr,sc)] = 0
        while q:
            r,c = q.popleft()
            for dr,dc in DIRS:
                rr = r+dr
                cc = c+dc
                if GRID[rr][cc] == "#":
                    continue
                if (rr,cc) in DIST:
                    continue
                DIST[(rr,cc)] = DIST[(r,c)] + 1
                q.append((rr,cc))
        return DIST

    DIST = runOne()

    ans = 0
    for r1 in range(0,R):
        for c1 in range(0,C):
            if GRID[r1][c1] == "#":
                continue
            for r2 in range(max(0, r1-2), min(R,r1+2+1)):
                for c2 in range(max(0, c1-2), min(C,c1+2+1)):
                    if GRID[r2][c2] == "#":
                        continue
                    oneNorm = abs(r2-r1) + abs(c2-c1)
                    if oneNorm > 2:
                        continue
                    if DIST[(r2,c2)] - DIST[(r1,c1)] - oneNorm >= 100:
                        ans += 1
    print(ans)

def doPart2():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    GRID = open(filename).read().strip().split("\n")

    R = len(GRID)
    C = len(GRID[0])

    for r in range(R):
        for c in range(C):
            if GRID[r][c] == "S":
                sr,sc = r,c

    def runOne():
        DIST = {}
        q=deque()
        q.append((sr,sc))
        DIST[(sr,sc)] = 0
        while q:
            r,c = q.popleft()
            for dr,dc in DIRS:
                rr = r+dr
                cc = c+dc
                if GRID[rr][cc] == "#":
                    continue
                if (rr,cc) in DIST:
                    continue
                DIST[(rr,cc)] = DIST[(r,c)] + 1
                q.append((rr,cc))
        return DIST

    DIST = runOne()

    ans = 0
    for r1 in range(0,R):
        for c1 in range(0,C):
            if GRID[r1][c1] == "#":
                continue
            for r2 in range(max(0, r1-20), min(R,r1+20+1)):
                for c2 in range(max(0, c1-20), min(C,c1+20+1)):
                    if GRID[r2][c2] == "#":
                        continue
                    oneNorm = abs(r2-r1) + abs(c2-c1)
                    if oneNorm > 20:
                        continue
                    if DIST[(r2,c2)] - DIST[(r1,c1)] - oneNorm >= 100:
                        ans += 1
    print(ans)

def main():
    # doPart1()
    doPart1Better()
    doPart2()

if __name__ == "__main__":
    main()

File: 2024/20/main.cpp

#include "../helpers/extras.hh"

std::vector<std::vector<int>> solveMaze(int sr, int sc, std::vector<std::string>& GRID)
{
    std::vector<std::vector<int>> DIST(GRID.size(), std::vector<int>(GRID[0].size(), -1));
    std::queue<std::array<int, 2>> queue;
    queue.push({sr, sc});
    DIST[sr][sc] = 0;
    while (!queue.empty()) {
        auto [r, c] = queue.front();
        queue.pop();
        for (auto [dr, dc] : DIRS) {
            int rr = r + dr, cc = c + dc;
            if (GRID[rr][cc] == '#') continue;
            if (DIST[rr][cc] != -1) continue;
            DIST[rr][cc] = DIST[r][c] + 1;
            queue.push({rr, cc});
        }
    }
    return DIST;
}

void doPart1(int sr, int sc, std::vector<std::string>& GRID, int cheatStep)
{
    std::vector<std::vector<int>> DIST = solveMaze(sr, sc, GRID);

    int R = GRID.size();
    int C = GRID[0].size();

    int minTime = R > 20 ? 100 : 50;

    long ans = 0;
    long best = 0;
    for (int r1 = 0; r1 < R; r1++) {
        for (int c1 = 0; c1 < C; c1++) {
            if (GRID[r1][c1] == '#') continue;
            for (int r2 = std::max(0, r1-cheatStep); r2 < std::min(R, r1+cheatStep+1); r2++) {
                for (int c2 = std::max(0, c1-cheatStep); c2 < std::min(C, c1+cheatStep+1); c2++) {
                    if (GRID[r2][c2] == '#') continue;
                    int oneNorm = std::abs(r2-r1) + std::abs(c2-c1);
                    if (oneNorm > cheatStep) continue;
                    long timeSaved = DIST[r2][c2] - DIST[r1][c1] - oneNorm;
                    if (timeSaved >= minTime) {
                        ans += 1;
                        if (timeSaved > best) {
                            best = timeSaved;
                        }
                    }
                }
            }
        }
    }
    printf("%ld, %ld\n", ans, best);
}

void doPart2(int sr, int sc, std::vector<std::string>& GRID, int cheatStep)
{
    std::vector<std::vector<int>> DIST = solveMaze(sr, sc, GRID);

    int R = GRID.size();
    int C = GRID[0].size();

    int minTime = R > 20 ? 100 : 50;

    long ans = 0;
    long best = 0;
    for (int r1 = 0; r1 < R; r1++) {
        for (int c1 = 0; c1 < C; c1++) {
            if (GRID[r1][c1] == '#') continue;
            for (int r2 = std::max(0, r1-cheatStep); r2 < std::min(R, r1+cheatStep+1); r2++) {
                for (int c2 = std::max(0, c1-cheatStep); c2 < std::min(C, c1+cheatStep+1); c2++) {
                    if (GRID[r2][c2] == '#') continue;
                    int oneNorm = std::abs(r2-r1) + std::abs(c2-c1);
                    if (oneNorm > cheatStep) continue;
                    long timeSaved = DIST[r2][c2] - DIST[r1][c1] - oneNorm;
                    if (timeSaved >= minTime) {
                        ans += 1;
                        if (timeSaved > best) {
                            best = timeSaved;
                        }
                    }
                }
            }
        }
    }
    printf("%ld, %ld\n", ans, best);
}

int main(int narg, char* args[])
{
    std::string filename = "example.dat";
    if (narg > 1) filename = std::string(args[1]);

    std::ifstream file(filename);
    std::string line;
    std::vector<std::string> GRID;
    while (getline(file, line)) {
        stripString(line);
        GRID.push_back(line);
    }
    file.close();

    int R = GRID.size();
    int C = GRID[0].size();

    int sr = -1, sc = -1;
    for (int i = 0; i < R; ++i) {
        for (int j = 0; j < C; ++j) {
            if (GRID[i][j] == 'S') sr = i, sc = j;
        }
    }

    doPart1(sr,sc,GRID,2);
    doPart2(sr,sc,GRID,20);
    return 0;
}
